File: fastapiApp/services/temperature.py
def convert_temperature(value: float, from_unit: str, to_unit: str) -> float:
    if from_unit == to_unit:
        return value
    # Convert input to Celsius
    if from_unit == "Celsius":
        celsius = value
    elif from_unit == "Fahrenheit":
        celsius = (value - 32) * 5/9
    elif from_unit == "Kelvin":
        celsius = value - 273.15
    else:
        raise ValueError("Invalid unit")

    # Convert Celsius to target
    if to_unit == "Celsius":
        return celsius
    elif to_unit == "Fahrenheit":
        return (celsius * 9/5) + 32
    elif to_unit == "Kelvin":
        return celsius + 273.15

File: golang/handlers/temperature.go
package handlers

import (
	"html/template"
	"main/services"
	"net/http"
	"strconv"
)

var tempPageTemplate = template.Must(template.ParseFiles("templates/temperature.html", "templates/base.html"))

type TempPageData struct {
	Value     float64
	FromUnit  string
	ToUnit    string
	Result    float64
	ErrorMsg  string
	Submitted bool
}

func TemperatureHandler(w http.ResponseWriter, r *http.Request) {
	data := TempPageData{}

	if r.Method == http.MethodPost {
		r.ParseForm()
		valueStr := r.FormValue("value")
		fromUnit := r.FormValue("from")
		toUnit := r.FormValue("to")

		val, err := strconv.ParseFloat(valueStr, 64)
		if err != nil {
			data.ErrorMsg = "Invalid input value"
		} else {
			result, convErr := services.ConvertTemperature(val, fromUnit, toUnit)
			if convErr != nil {
				data.ErrorMsg = convErr.Error()
			} else {
				data.Result = result
				data.Value = val
				data.FromUnit = fromUnit
				data.ToUnit = toUnit
				data.Submitted = true
			}
		}
	}

	tempPageTemplate.Execute(w, data)
}
